packages/dev/core/src/Maths/math.vector.ts

```diff
@@ -1150,22 +1150,43 @@ export class Vector3 {
      * @param result defines the Vector3 where to store the result
      */
     public projectOnPlaneToRef(plane: Plane, origin: Vector3, result: Vector3): void {
-        const n = plane.normal;
-        const d = plane.d;
-
-        const V = MathTmp.Vector3[0];
-
-        // ray direction
-        this.subtractToRef(origin, V);
+        // Use the normal scaled to the plane offset as the origin for the formula
+        const planeOrigin = MathTmp.Vector3[0];
+        plane.normal.scaleToRef(-plane.d, planeOrigin)
+
+        // Since the normal in Babylon should point toward the viewer, invert it for the dot product
+        const inverseNormal = MathTmp.Vector3[1];
+        plane.normal.negateToRef(inverseNormal);
+        
+        // This vector is the direction
+        const direction = this;
+
+        // Set a margin to prevent issues with directions very close to perpendicular to the plane
+        const epsilon = 0.000001;

```

Comment:
This exists in math.constants (and is already imported to the math module)

Reply:
Good call! It's probably better to use a consistent fudge value. :)